File: go_app/internal/ui/card.go
package ui

import (
	"fmt"
	"strings"
)

// Card represents a card component
type Card struct {
	CardOptions   map[string]interface{}
	HeaderOptions map[string]interface{}
	BodyOptions   map[string]interface{}
	FooterOptions map[string]interface{}
	Card          map[string]interface{}
	Header        string
	Body          string
	Footer        string
}

// NewCard creates a new card with the given parameters
func NewCard(card map[string]interface{}, cardOptions map[string]interface{}, headerOptions map[string]interface{}, bodyOptions map[string]interface{}, footerOptions map[string]interface{}, header string, body string, footer string) *Card {
	if card == nil {
		card = make(map[string]interface{})
	}
	if cardOptions == nil {
		cardOptions = make(map[string]interface{})
	}
	if headerOptions == nil {
		headerOptions = make(map[string]interface{})
	}
	if bodyOptions == nil {
		bodyOptions = make(map[string]interface{})
	}
	if footerOptions == nil {
		footerOptions = make(map[string]interface{})
	}

	return &Card{
		CardOptions:   cardOptions,
		HeaderOptions: headerOptions,
		BodyOptions:   bodyOptions,
		FooterOptions: footerOptions,
		Card:          card,
		Header:        header,
		Body:          body,
		Footer:        footer,
	}
}

// HasHeader returns whether the card has a header
func (c *Card) HasHeader() bool {
	return c.Header != "" || (c.Card["header"] != nil && c.Card["header"] != "")
}

// GetHeader returns the header content
func (c *Card) GetHeader() string {
	if c.Header != "" {
		return c.Header
	}
	if header, ok := c.Card["header"].(string); ok {
		return header
	}
	return ""
}

// HasFooter returns whether the card has a footer
func (c *Card) HasFooter() bool {
	return c.Footer != "" || (c.Card["footer"] != nil && c.Card["footer"] != "")
}

// GetFooter returns the footer content
func (c *Card) GetFooter() string {
	if c.Footer != "" {
		return c.Footer
	}
	if footer, ok := c.Card["footer"].(string); ok {
		return footer
	}
	return ""
}

// GetBody returns the body content
func (c *Card) GetBody() string {
	if c.Body != "" {
		return c.Body
	}
	if body, ok := c.Card["body"].(string); ok {
		return body
	}
	return ""
}

// Render generates the HTML for the card
func (c *Card) Render() string {
	var parts []string

	// Add card class to card options
	if _, ok := c.CardOptions["class"]; !ok {
		c.CardOptions["class"] = "gl-card"
	} else {
		class := c.CardOptions["class"].(string)
		if !strings.Contains(class, "gl-card") {
			c.CardOptions["class"] = fmt.Sprintf("gl-card %s", class)
		}
	}

	// Start card div
	parts = append(parts, fmt.Sprintf(`<div %s>`, formatAttributes(c.CardOptions)))

	// Add header if present
	if c.HasHeader() {
		// Add header class to header options
		if _, ok := c.HeaderOptions["class"]; !ok {
			c.HeaderOptions["class"] = "gl-card-header"
		} else {
			class := c.HeaderOptions["class"].(string)
			if !strings.Contains(class, "gl-card-header") {
				c.HeaderOptions["class"] = fmt.Sprintf("gl-card-header %s", class)
			}
		}
		parts = append(parts, fmt.Sprintf(`<div %s>%s</div>`, formatAttributes(c.HeaderOptions), c.GetHeader()))
	}

	// Add body
	// Add body class to body options
	if _, ok := c.BodyOptions["class"]; !ok {
		c.BodyOptions["class"] = "gl-card-body"
	} else {
		class := c.BodyOptions["class"].(string)
		if !strings.Contains(class, "gl-card-body") {
			c.BodyOptions["class"] = fmt.Sprintf("gl-card-body %s", class)
		}
	}
	parts = append(parts, fmt.Sprintf(`<div %s>%s</div>`, formatAttributes(c.BodyOptions), c.GetBody()))

	// Add footer if present
	if c.HasFooter() {
		// Add footer class to footer options
		if _, ok := c.FooterOptions["class"]; !ok {
			c.FooterOptions["class"] = "gl-card-footer"
		} else {
			class := c.FooterOptions["class"].(string)
			if !strings.Contains(class, "gl-card-footer") {
				c.FooterOptions["class"] = fmt.Sprintf("gl-card-footer %s", class)
			}
		}
		parts = append(parts, fmt.Sprintf(`<div %s>%s</div>`, formatAttributes(c.FooterOptions), c.GetFooter()))
	}

	// Close card div
	parts = append(parts, "</div>")

	return strings.Join(parts, "\n")
}

File: rust_app/src/components/pajamas/card.rs
use std::collections::HashMap;

/// Renders a Pajamas compliant card element
#[derive(Debug, Clone)]
pub struct CardComponent {
    /// Options for the card
    card_options: HashMap<String, String>,
    /// Options for the header
    header_options: HashMap<String, String>,
    /// Options for the body
    body_options: HashMap<String, String>,
    /// Options for the footer
    footer_options: HashMap<String, String>,
    /// The card structure
    card: HashMap<String, String>,
    /// The header content slot
    header_slot: Option<String>,
    /// The body content slot
    body_slot: Option<String>,
    /// The footer content slot
    footer_slot: Option<String>,
}

impl CardComponent {
    /// Create a new card component
    pub fn new(
        card: Option<HashMap<String, String>>,
        card_options: Option<HashMap<String, String>>,
        header_options: Option<HashMap<String, String>>,
        body_options: Option<HashMap<String, String>>,
        footer_options: Option<HashMap<String, String>>,
    ) -> Self {
        Self {
            card_options: card_options.unwrap_or_default(),
            header_options: header_options.unwrap_or_default(),
            body_options: body_options.unwrap_or_default(),
            footer_options: footer_options.unwrap_or_default(),
            card: card.unwrap_or_default(),
            header_slot: None,
            body_slot: None,
            footer_slot: None,
        }
    }

    /// Set the header content slot
    pub fn with_header_slot(mut self, header: String) -> Self {
        self.header_slot = Some(header);
        self
    }

    /// Set the body content slot
    pub fn with_body_slot(mut self, body: String) -> Self {
        self.body_slot = Some(body);
        self
    }

    /// Set the footer content slot
    pub fn with_footer_slot(mut self, footer: String) -> Self {
        self.footer_slot = Some(footer);
        self
    }

    /// Check if the header is present
    fn has_header(&self) -> bool {
        self.header_slot.is_some() || self.card.contains_key("header")
    }

    /// Get the header content
    fn get_header(&self) -> Option<String> {
        self.header_slot
            .clone()
            .or_else(|| self.card.get("header").cloned())
    }

    /// Get the body content
    fn get_body(&self) -> Option<String> {
        self.body_slot
            .clone()
            .or_else(|| self.card.get("body").cloned())
    }

    /// Check if the footer is present
    fn has_footer(&self) -> bool {
        self.footer_slot.is_some() || self.card.contains_key("footer")
    }

    /// Get the footer content
    fn get_footer(&self) -> Option<String> {
        self.footer_slot
            .clone()
            .or_else(|| self.card.get("footer").cloned())
    }

    /// Format the card options
    fn formatted_card_options(&self) -> String {
        self.card_options
            .iter()
            .map(|(key, value)| format!(" {}=\"{}\"", key, value))
            .collect::<Vec<String>>()
            .join("")
    }

    /// Format the header options
    fn formatted_header_options(&self) -> String {
        self.header_options
            .iter()
            .map(|(key, value)| format!(" {}=\"{}\"", key, value))
            .collect::<Vec<String>>()
            .join("")
    }

    /// Format the body options
    fn formatted_body_options(&self) -> String {
        self.body_options
            .iter()
            .map(|(key, value)| format!(" {}=\"{}\"", key, value))
            .collect::<Vec<String>>()
            .join("")
    }

    /// Format the footer options
    fn formatted_footer_options(&self) -> String {
        self.footer_options
            .iter()
            .map(|(key, value)| format!(" {}=\"{}\"", key, value))
            .collect::<Vec<String>>()
            .join("")
    }

    /// Render the card component
    pub fn render(&self) -> String {
        let mut html = String::new();
        let card_options = self.formatted_card_options();

        // Card container
        html.push_str(&format!(r#"<div class="gl-card"{}>"#, card_options));

        // Header
        if self.has_header() {
            let header_options = self.formatted_header_options();
            html.push_str(&format!(
                r#"<div class="gl-card-header"{}>"#,
                header_options
            ));

            if let Some(header) = self.get_header() {
                html.push_str(&header);
            }

            html.push_str("</div>");
        }

        // Body
        let body_options = self.formatted_body_options();
        html.push_str(&format!(r#"<div class="gl-card-body"{}>"#, body_options));

        if let Some(body) = self.get_body() {
            html.push_str(&body);
        }

        html.push_str("</div>");

        // Footer
        if self.has_footer() {
            let footer_options = self.formatted_footer_options();
            html.push_str(&format!(
                r#"<div class="gl-card-footer"{}>"#,
                footer_options
            ));

            if let Some(footer) = self.get_footer() {
                html.push_str(&footer);
            }

            html.push_str("</div>");
        }

        html.push_str("</div>");
        html
    }
}
